### PR TITLE
Add new timelist view and plugin

### PR DESCRIPTION
**Pull request by shefalijoshi**
Resolves: #4765
Adds new list component
Adds new timelist view and domain object type

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?
* [x] Is this change backwards compatible? For example, developers won't need to change how they are calling the API or how they've extended core plugins such as Tables or Plots.

### Author Checklist

* [x] Changes address original issue?
* [x] Unit tests included and/or updated with changes?
* [x] Command line build passes?
* [x] Has this been smoke tested?
* [x] Testing instructions included in associated issue?

### Reviewer Checklist

* [x] Changes appear to address issue?
* [x] Changes appear not to be breaking changes?
* [x] Appropriate unit tests included?
* [x] Code style and in-line documentation are appropriate?
* [x] Commit messages meet standards?
* [ ] Has associated issue been labelled unverified? (only applicable if this PR closes the issue)
* [ ] Has associated issue been labelled bug? (only applicable if this PR is for a bug fix)

### CODE REVIEW

**Review by charlesh88 — CHANGES_REQUESTED** [review]
- [x] While editing, the view should not filter by the Inspector show settings.
- [x] Need to add string filter.
- [ ] @charlesh88 Need to add design for error state for string filter
image

- [ ] We should be able to drag a Plan from the tree to populate in addition to importing a JSON file.
- [x] Future activities need to be classed as such per row.
- [x] We _must_ have sticky headers, only the body should scroll, and we'd also benefit from value-based column sizing. We should use the same table approach we use elsewhere.
- [x] Inspector editing UI is showing in browse mode, but shouldn't.
- [ ] @charlesh88 to polish Inspector UI and list view styles.
- [ ] @charlesh88 to remove hover class on row.

**Comment by charlesh88** [comment]
Have pushed changes to the branch, here's what it looks like:
Screen Shot 2022-02-04 at 4 57 17 PM
Screen Shot 2022-02-04 at 4 57 25 PM

Also created a spreadsheet that generates valid Plan files: https://docs.google.com/spreadsheets/d/1Hq1tZwK0QpsA6Q9l7hNOr0kgZvei0-5e6HQcG8V-0jQ/edit#gid=1940164110

Screen Shot 2022-02-04 at 4 57 34 PM

Example output JSON:
[plan for Timelist view CH.json.zip](https://github.com/nasa/openmct/files/8007041/plan.for.Timelist.view.CH.json.zip)

@charlesh88 TODOs:
- [ ] Complete Inspector layout;

**Comment by charlesh88** [comment]
@shefalijoshi Nice work on the headers and Inspector browse view. I pushed lots of improvements and added a comment about filtering to the requested changes - have a look.

File: src/plugins/timelist/Timelist.vue [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+            if (this.showAll === true) {
+                return true;
+            }
+
+            //current event or future start event or past end event
+            const isCurrent = (this.noCurrent === false && this.timestamp >= activity.start && this.timestamp <= activity.end);
+            const isPast = (this.timestamp > activity.end && (this.viewBounds.pastEnd === undefined || activity.end >= this.viewBounds.pastEnd(this.timestamp)));
+            const isFuture = (this.timestamp < activity.start && (this.viewBounds.futureStart === undefined || activity.start <= this.viewBounds.futureStart(this.timestamp)));
+
+            return isCurrent || isPast || isFuture;
```
Comment (nikhilmandlik) [inline src/plugins/timelist/Timelist.vue:269]: hmm, interesting check, any activity has to be either current, past or future 😅 

I think can divide into `const isPast= this.timestamp > activity.end` and condition into separate variable
then can use like
`return isCurrent || (isPast && condition1) || (isFuture && condition2);`

File: src/plugins/timelist/pluginSpec.js [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+            const applicableViews = openmct.objectViews.get(timelistDomainObject, [timelistDomainObject]);
+            timelistView = applicableViews.find((viewProvider) => viewProvider.key === 'timelist.view');
+            let view = timelistView.view(timelistDomainObject, []);
+            view.show(child, true);
+
+            return Vue.nextTick();
+        });
+
+        it('displays the activities', () => {
+            const items = element.querySelectorAll('.c-table__body .c-list-item');
```
Comment (nikhilmandlik) [inline src/plugins/timelist/pluginSpec.js:154]: need to use , `.js-xxx` classes for targeting elements from js file , I also need to do similar changes in my test 😜

File: src/plugins/timelist/inspector/EventProperties.vue [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+         class="c-inspect-properties__value"
+    >
+        {{ activitiesOptions[index] }} <span v-if="index > 1">{{ duration }} {{ durationOptions[durationIndex] }}</span>
+    </div>
+</li>
+</template>
+
+<script>
+const ACTIVITIES_OPTIONS = [
+    'Don\'t show',
```
[the remark this answers is not in the document]
Comment (charlesh88) [reply src/plugins/timelist/inspector/EventProperties.vue:43]: @nikhilmandlik What do you mean? The escaped single quote?

**Comment by charlesh88** [comment]
@shefalijoshi I've just pushed a small tweak to the Inspector filter input. Is looking very good, the only thing I see missing now is auto-scrolling the view. Where are we on that?

File: src/plugins/timelist/constants.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
+export const SORT_ORDER_OPTIONS = [
+    {
+        label: 'Start Ascending',
```
[the remark this answers is not in the document]
Comment (shefalijoshi) [reply src/plugins/timelist/constants.js:3]: @charlesh88 Do you have a preference on what these labels should be?

File: src/plugins/timelist/Timelist.vue [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+            const direction = sortOrder.direction.toLowerCase() === 'asc';
+            this.defaultSort = {
+                property,
+                defaultDirection: direction
+            };
+        },
+        sortByStartTime(a, b) {
+            const numA = parseInt(a.start, 10);
+            const numB = parseInt(b.start, 10);
+            if (numA > numB) {
```
[the remark this answers is not in the document]
Comment (shefalijoshi) [reply src/plugins/timelist/Timelist.vue:332]: Done

File: src/plugins/timelist/TimelistViewProvider.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import Vue from 'vue';
+
+export default function TimelistViewProvider(openmct) {
+
+    return {
+        key: 'timelist.view',
+        name: 'Time List',
+        cssClass: 'icon-timelist',
+        canView(domainObject) {
+            return domainObject.type === 'timelist';
```
[the remark this answers is not in the document]
Comment (shefalijoshi) [reply src/plugins/timelist/TimelistViewProvider.js:33]: done

File: src/plugins/timelist/inspector/EventProperties.vue (hunk 3 above)
[the remark this answers is not in the document]
Comment (shefalijoshi) [reply src/plugins/timelist/inspector/EventProperties.vue:43]: Not a typo

File: src/plugins/timelist/Timelist.vue [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+
+                return;
+            }
+
+            this.showAll = false;
+
+            this.viewBounds = {};
+
+            if (currentEventsIndex === 0) {
+                this.noCurrent = true;
```
[the remark this answers is not in the document]
Comment (shefalijoshi) [reply src/plugins/timelist/Timelist.vue:193]: done

File: src/plugins/timelist/inspector/TimelistPropertiesView.vue [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+            <filtering @updated="eventPropertiesUpdated" />
+        </ul>
+    </div>
+</div>
+</template>
+
+<script>
+
+import EventProperties from './EventProperties.vue';
+import { SORT_ORDER_OPTIONS } from "../constants";
```
[the remark this answers is not in the document]
Comment (shefalijoshi) [reply src/plugins/timelist/inspector/TimelistPropertiesView.vue:77]: done

File: src/plugins/timelist/pluginSpec.js (hunk 2 above)
Comment (shefalijoshi) [reply src/plugins/timelist/pluginSpec.js]: Done

**Comment by jvigliotta** [comment]
@shefalijoshi what's the status of this?

**Comment by charlesh88** [comment]
My CSS for past/current/future events has been wiped out. Investigating...

**Comment by charlesh88** [comment]
I made and pushed some tweaks to where and how styles were defined for this, as well as the sorting menu text.

**Review by charlesh88 — APPROVED** [review]
LGTM, nice work!

**Comment by shefalijoshi** [comment]
Correlation logic of data for endpoints included in an X-Y plot should be removed from Open MCT and included as a plugin if necessary.

**Review by nikhilmandlik — DISMISSED** [review]
LGTM

**Comment by shefalijoshi** [comment]
@nikhilmandlik Please also merge the corresponding VIPER-OPENMCT PR to install the timelist plugin

**Review by nikhilmandlik — APPROVED** [review]
LGTM

**Pull request merged**